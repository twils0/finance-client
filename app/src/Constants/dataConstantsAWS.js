// eslint-disable-next-line
export const actionTypes = {
  SET_AWS_STATUS: 'DATA_AUTH_AWS_STATUS',
  SET_AWS_USER: 'DATA_AUTH_SET_AWS_USER',
  RESET_AWS_STATE: 'DATA_AUTH_RESET_AWS_STATE',
};
