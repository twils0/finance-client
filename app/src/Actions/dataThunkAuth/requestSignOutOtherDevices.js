import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import { setAuthStatus } from '../dataActionsAuth';
import { setAWSStatus, setAWSUser } from '../dataActionsAWS';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';

const requestSignOutOtherDevices = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'email')) {
    throw new Error(`Please enter a value for the 'password' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'password')) {
    throw new Error(`Please enter a value for the 'password' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'remembered')) {
    throw new Error(`Please enter a value for the 'remembered' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    let state = getState();

    if (
      state.data.auth.status[statusNames.SIGN_OUT_DEVICES].status !== requestStatusTypes.LOADING
    ) {
      dispatch(setAuthStatus({
        id: statusNames.SIGN_OUT_DEVICES,
        status: requestStatusTypes.LOADING,
      }));

      let { user } = state.data.aws;

      try {
        if (!user) {
          ({ user } = await dispatch(requestAWSUser()));
        }

        await new Promise((resolve, reject) => {
          user.setDeviceStatusRemembered({
            onSuccess: () => {
              resolve();
            },
            onFailure: (error) => {
              reject(error);
            },
          });
        });

        await new Promise((resolve, reject) => {
          user.globalSignOut({
            onSuccess: () => {
              resolve();
            },
            onFailure: (error) => {
              reject(error);
            },
          });
        });

        dispatch(setAWSStatus({ status: requestStatusTypes.LOADING }));

        user = await Auth.signIn(payload.email, payload.password);

        dispatch(setAWSUser({ user }));
        dispatch(setAWSStatus({ status: requestStatusTypes.SUCCESS }));

        if (!payload.remembered) {
          await new Promise((resolve, reject) => {
            user.setDeviceStatusNotRemembered({
              onSuccess: () => {
                resolve();
              },
              onFailure: (error) => {
                reject(error);
              },
            });
          });
        }
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);
        state = getState();

        dispatch(setAuthStatus({
          id: statusNames.SIGN_OUT_DEVICES,
          status: requestStatusTypes.ERROR,
        }));
        if (state.data.aws.status === requestStatusTypes.LOADING) {
          dispatch(setAWSStatus({ status: requestStatusTypes.ERROR }));
        }

        raven.captureException(error, {
          logger: 'requestSignOutOtherDevices',
        });

        return Promise.reject(error);
      }

      dispatch(setAuthStatus({
        id: statusNames.SIGN_OUT_DEVICES,
        status: requestStatusTypes.SUCCESS,
      }));

      return user;
    }

    return null;
  };
};

export default requestSignOutOtherDevices;
