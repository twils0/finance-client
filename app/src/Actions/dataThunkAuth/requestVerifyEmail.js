import axios from 'axios';
import raven from 'raven-js';

import { requestStatusTypes, URLs, axiosConfig } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import requestLogout from './requestLogout';
import { setAuthStatus } from '../dataActionsAuth';

import handleErrorCatch from '../../handleErrorCatch';

import { demo } from '../../../../mode.config.json';

const requestVerifyEmail = () => async (dispatch, getState) => {
  const state = getState();

  if (state.data.auth.status[statusNames.VERIFY_EMAIL].status !== requestStatusTypes.LOADING) {
    dispatch(
      setAuthStatus({
        id: statusNames.VERIFY_EMAIL,
        status: requestStatusTypes.LOADING,
      }),
    );

    let { user } = state.data.aws;

    try {
      if (!user) {
        ({ user } = await dispatch(requestAWSUser()));
      }

      if (!demo) {
        const idToken = user.signInUserSession.idToken.jwtToken;
        const accessToken = user.signInUserSession.accessToken.jwtToken;

        await axios.post(
          URLs.EMAILS,
          {},
          {
            headers: {
              Authorization: idToken,
            },
            params: {
              accessToken,
            },
            ...axiosConfig.DB,
          },
        );
      }
    } catch (errorCatch) {
      const error = handleErrorCatch(errorCatch);

      dispatch(
        setAuthStatus({
          id: statusNames.VERIFY_EMAIL,
          status: requestStatusTypes.ERROR,
        }),
      );

      if (typeof error === 'object' && error && error.code === 'NotAuthorizedException') {
        dispatch(requestLogout());

        return null;
      }

      if (!demo) {
        raven.captureException(error, {
          logger: 'requestVerifyEmail',
        });
      }

      return Promise.reject(error);
    }

    dispatch(
      setAuthStatus({
        id: statusNames.VERIFY_EMAIL,
        status: requestStatusTypes.SUCCESS,
      }),
    );
  }

  return null;
};

export default requestVerifyEmail;
