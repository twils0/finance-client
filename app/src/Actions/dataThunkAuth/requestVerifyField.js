import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import { requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames } from '../../Constants/dataConstantsAuth';
import requestLogout from './requestLogout';
import requestAWSUser from '../dataThunkAWS/requestAWSUser';
import { setAuthStatus } from '../dataActionsAuth';

import handleErrorCatch from '../../handleErrorCatch';

const requestVerifyField = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'field')) {
    throw new Error(`Please enter a value for the 'field' key - ${JSON.stringify(payload)}`);
  }
  if (payload.field !== 'phone_number') {
    throw new Error(
      `Please enter 'phone_number' as the value for the 'field' key - ${JSON.stringify(payload)}`,
    );
  }

  return async (dispatch, getState) => {
    const state = getState();

    if (state.data.auth.status[statusNames.VERIFY_PHONE].status !== requestStatusTypes.LOADING) {
      switch (payload.field) {
        case 'phone_number':
          dispatch(
            setAuthStatus({
              id: statusNames.VERIFY_PHONE,
              status: requestStatusTypes.LOADING,
            }),
          );
          break;
        default:
          break;
      }

      let { user } = state.data.aws;

      try {
        if (!user) {
          ({ user } = await dispatch(requestAWSUser()));
        }

        await Auth.verifyUserAttribute(user, payload.field);
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);

        switch (payload.field) {
          case 'phone_number':
            dispatch(
              setAuthStatus({
                id: statusNames.VERIFY_PHONE,
                status: requestStatusTypes.ERROR,
              }),
            );
            break;
          default:
            break;
        }

        if (typeof error === 'object' && error && error.code === 'NotAuthorizedException') {
          dispatch(requestLogout());

          return null;
        }

        raven.captureException(error, {
          logger: 'requestVerifyField',
        });

        return Promise.reject({ field: payload.field, error });
      }

      switch (payload.field) {
        case 'phone_number':
          dispatch(
            setAuthStatus({
              id: statusNames.VERIFY_PHONE,
              status: requestStatusTypes.SUCCESS,
            }),
          );
          break;
        default:
          break;
      }
    }

    return null;
  };
};

export default requestVerifyField;
