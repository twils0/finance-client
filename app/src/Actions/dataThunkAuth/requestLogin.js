import { Auth } from 'aws-amplify';
import raven from 'raven-js';

import handleErrorCatch from '../../handleErrorCatch';
import { requestStatusTypes } from '../../Constants/universalConstants';
import { statusNames, codeTypeNames } from '../../Constants/dataConstantsAuth';
import { setAuthStatus, setAuthenticated } from '../dataActionsAuth';
import { setAWSStatus, setAWSUser } from '../dataActionsAWS';
import requestSignOutOtherDevices from './requestSignOutOtherDevices';
import loadAWSFields from '../dataThunkAccount/loadAWSFields';
import loadSecurities from '../dataThunkWatchlist/loadSecurities';
import requestVerifyField from './requestVerifyField';
import requestVerifyEmail from './requestVerifyEmail';
import { fieldNames } from '../../Constants/dataConstantsAccount';

import { demo } from '../../../../mode.config.json';

const requestLogin = (payload) => {
  if (!Object.prototype.hasOwnProperty.call(payload, 'email')) {
    throw new Error(`Please enter a value for the 'email' key - ${JSON.stringify(payload)}`);
  }
  if (!Object.prototype.hasOwnProperty.call(payload, 'password')) {
    throw new Error(`Please enter a value for the 'password' key - ${JSON.stringify(payload)}`);
  }

  return async (dispatch, getState) => {
    let state = getState();

    if (state.data.auth.status[statusNames.LOGIN].status !== requestStatusTypes.LOADING) {
      dispatch(setAuthStatus({ id: statusNames.LOGIN, status: requestStatusTypes.LOADING }));
      dispatch(setAWSStatus({ status: requestStatusTypes.LOADING }));

      let user = null;

      try {
        if (!demo) {
          raven.setUserContext({
            email: payload.email,
          });

          user = await Auth.signIn(payload.email, payload.password);

          dispatch(setAWSUser({ user }));
          dispatch(setAWSStatus({ status: requestStatusTypes.SUCCESS }));
        }
      } catch (errorCatch) {
        const error = handleErrorCatch(errorCatch);

        dispatch(
          setAuthStatus({
            id: statusNames.LOGIN,
            status: requestStatusTypes.ERROR,
          }),
        );
        dispatch(setAWSStatus({ status: requestStatusTypes.ERROR }));

        if (
          typeof error === 'object'
          && error
          && (error.code === 'LimitExceededException'
            || error.code === 'UserNotFoundException'
            || (error.code === 'NotAuthorizedException'
              && error.message === 'Incorrect username or password.')
            || error.code === 'UserNotConfirmedException')
        ) {
          return Promise.reject(error);
        }

        if (!demo) {
          raven.captureException(error, {
            logger: 'requestLogin',
          });
        }

        return Promise.reject(error);
      }

      let mfa = true;

      if (!demo) {
        if (user.challengeName !== 'SMS_MFA') {
          mfa = false;

          try {
            await dispatch(requestSignOutOtherDevices({ ...payload, remembered: true }));

            dispatch(loadSecurities());

            await dispatch(loadAWSFields());

            state = getState();

            if (state.data.auth.codeTypes[codeTypeNames.VERIFY_PHONE].needed) {
              await dispatch(requestVerifyField({ field: 'phone_number' }));
            } else if (
              state.data.auth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed
              || (state.data.auth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed
                && state.data.account.fields[fieldNames.EMAIL_ADDITIONAL].value)
            ) {
              await dispatch(requestVerifyEmail());
            }

            if (
              !state.data.auth.codeTypes[codeTypeNames.VERIFY_PHONE].needed
              && !state.data.auth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed
              && (!state.data.auth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed
                || !state.data.account.fields[fieldNames.EMAIL_ADDITIONAL].value)
            ) {
              dispatch(setAuthenticated({ authenticated: true }));
            }
          } catch (errorCatch) {
            const error = handleErrorCatch(errorCatch);

            dispatch(
              setAuthStatus({
                id: statusNames.LOGIN,
                status: requestStatusTypes.ERROR,
              }),
            );

            if (!demo) {
              raven.captureException(error, {
                logger: 'requestLogin',
              });
            }

            return Promise.reject(error);
          }
        }
      }

      dispatch(
        setAuthStatus({
          id: statusNames.LOGIN,
          status: requestStatusTypes.SUCCESS,
        }),
      );

      return mfa;
    }

    return null;
  };
};

export default requestLogin;
