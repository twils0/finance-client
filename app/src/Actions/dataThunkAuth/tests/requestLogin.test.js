/**
 * @jest-environment node
 */

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Auth } from 'aws-amplify';

import { initialState as initialStateAuth } from '../../../Reducers/dataReducersAuth';
import { requestStatusTypes } from '../../../Constants/universalConstants';
import {
  actionTypes as actionTypesAuth,
  codeTypeNames,
  statusNames,
} from '../../../Constants/dataConstantsAuth';
import { actionTypes as actionTypesAWS } from '../../../Constants/dataConstantsAWS';
import requestLogin from '../requestLogin';
import requestVerifyField from '../requestVerifyField';
import requestVerifyEmail from '../requestVerifyEmail';
import requestSignOutOtherDevices from '../requestSignOutOtherDevices';
import loadAWSFields from '../../dataThunkAccount/loadAWSFields';
import loadSecurities from '../../dataThunkWatchlist/loadSecurities';

jest.mock('aws-amplify', () => ({
  Auth: {
    signIn: jest.fn(),
  },
}));
jest.mock('../requestSignOutOtherDevices', () => jest.fn());
jest.mock('../../dataThunkAccount/loadAWSFields', () => jest.fn());
jest.mock('../../dataThunkWatchlist/loadSecurities', () => jest.fn());
jest.mock('../requestVerifyField', () => jest.fn());
jest.mock('../requestVerifyEmail', () => jest.fn());

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

const loginPayload = { email: 'testEmail', password: 'testPassword' };
const signOutPayload = { ...loginPayload, remembered: true };

describe('dataThunkAuth', () => {
  describe('requestLogin', () => {
    afterEach(() => {
      Auth.signIn.mockReset();
      requestSignOutOtherDevices.mockReset();
      loadSecurities.mockReset();
      loadAWSFields.mockReset();
      requestVerifyField.mockReset();
      requestVerifyEmail.mockReset();
    });

    it("fails and throws an error when missing 'email' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      Auth.signIn.mockReturnValue(Promise.resolve());

      const emptyPayload = {};

      try {
        await store.dispatch(requestLogin(emptyPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(Auth.signIn).not.toBeCalled();
      expect(requestSignOutOtherDevices).not.toBeCalled();
      expect(loadSecurities).not.toBeCalled();
      expect(loadAWSFields).not.toBeCalled();
      expect(requestVerifyField).not.toBeCalled();
    });

    it("fails and throws an error when missing 'password' key in payload", async () => {
      const store = mockStore({ data: { auth: initialStateAuth } });

      Auth.signIn.mockReturnValue(Promise.resolve());

      const wrongPayload = { email: 'testEmail' };

      try {
        await store.dispatch(requestLogin(wrongPayload));
      } catch (errorCatch) {
        expect(errorCatch).toMatchSnapshot();
      }

      expect(Auth.signIn).not.toBeCalled();
      expect(requestSignOutOtherDevices).not.toBeCalled();
      expect(loadSecurities).not.toBeCalled();
      expect(loadAWSFields).not.toBeCalled();
      expect(requestVerifyField).not.toBeCalled();
    });

    it('fails and returns promise reject when Auth.signIn throws an error', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_PHONE].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.ERROR,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.reject(error));

      try {
        await store.dispatch(requestLogin(loginPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).not.toBeCalled();
      expect(loadSecurities).not.toBeCalled();
      expect(loadAWSFields).not.toBeCalled();
      expect(requestVerifyField).not.toBeCalled();
    });

    it('fails and returns promise reject when requestVerifyField throws an error, when phone verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_PHONE].needed = true;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const user = { test: 'testUser' };
      const verifyPhonePayload = { field: 'phone_number' };
      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(user));
      requestSignOutOtherDevices.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      requestVerifyField.mockReturnValue(() => Promise.reject(error));

      try {
        await store.dispatch(requestLogin(loginPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).toBeCalledWith(signOutPayload);
      expect(loadSecurities).toBeCalled();
      expect(loadAWSFields).toBeCalled();
      expect(requestVerifyField).toBeCalledWith(verifyPhonePayload);
    });

    it('fails and returns promise reject when requestVerifyEmail throws an error, when email verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = true;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const user = { test: 'testUser' };
      const verifyPhonePayload = { field: 'phone_number' };
      const error = { code: 'testError', message: 'testMessage' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.ERROR,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(user));
      requestSignOutOtherDevices.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      requestVerifyField.mockReturnValue(() => Promise.reject(error));

      try {
        await store.dispatch(requestLogin(loginPayload));
      } catch (errorCatch) {
        expect(errorCatch).toEqual(error);
      }

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).toBeCalledWith(signOutPayload);
      expect(loadSecurities).toBeCalled();
      expect(loadAWSFields).toBeCalled();
      expect(requestVerifyField).toBeCalledWith(verifyPhonePayload);
    });

    it('creates the correct actions with the correct payloads when phone number verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const user = { test: 'testUser' };
      const verifyPhonePayload = { field: 'phone_number' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(user));
      requestSignOutOtherDevices.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      requestVerifyField.mockReturnValue(() => Promise.resolve());

      const result = await store.dispatch(requestLogin(loginPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(false);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).toBeCalledWith(signOutPayload);
      expect(loadSecurities).toBeCalled();
      expect(loadAWSFields).toBeCalled();
      expect(requestVerifyField).toBeCalledWith(verifyPhonePayload);
    });

    it('creates the correct actions with the correct payloads, when email verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_PHONE].needed = false;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const user = { test: 'testUser' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(user));
      requestSignOutOtherDevices.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      requestVerifyEmail.mockReturnValue(() => Promise.resolve());

      const result = await store.dispatch(requestLogin(loginPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(false);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).toBeCalledWith(signOutPayload);
      expect(loadSecurities).toBeCalled();
      expect(loadAWSFields).toBeCalled();
      expect(requestVerifyEmail).toBeCalled();
    });

    it('creates the correct actions with the correct payloads, when email additional verify needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_PHONE].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = true;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = true;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const user = { test: 'testUser' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(user));
      requestSignOutOtherDevices.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      requestVerifyEmail.mockReturnValue(() => Promise.resolve());

      const result = await store.dispatch(requestLogin(loginPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(false);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).toBeCalledWith(signOutPayload);
      expect(loadSecurities).toBeCalled();
      expect(loadAWSFields).toBeCalled();
      expect(requestVerifyEmail).toBeCalled();
    });

    it('creates the correct actions with the correct payloads when MFA needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_PHONE].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const user = { test: 'testUser', challengeName: 'SMS_MFA' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(user));

      const result = await store.dispatch(requestLogin(loginPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(true);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).not.toBeCalled();
      expect(loadAWSFields).not.toBeCalled();
      expect(loadSecurities).not.toBeCalled();
      expect(requestVerifyField).not.toBeCalled();
    });

    it('creates the correct actions with the correct payloads when MFA not needed', async () => {
      const stateBeforeAuth = JSON.parse(JSON.stringify(initialStateAuth));
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_PHONE].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL].needed = false;
      stateBeforeAuth.codeTypes[codeTypeNames.VERIFY_EMAIL_ADDITIONAL].needed = false;

      const store = mockStore({ data: { auth: stateBeforeAuth } });

      const user = { test: 'testUser' };
      const expectedActions = [
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.LOADING,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_USER,
          payload: {
            user,
          },
        },
        {
          type: actionTypesAWS.SET_AWS_STATUS,
          payload: {
            status: requestStatusTypes.SUCCESS,
          },
        },
        {
          type: actionTypesAuth.SET_AUTHENTICATED,
          payload: {
            authenticated: true,
          },
        },
        {
          type: actionTypesAuth.SET_AUTH_STATUS,
          payload: {
            id: statusNames.LOGIN,
            status: requestStatusTypes.SUCCESS,
          },
        },
      ];

      Auth.signIn.mockReturnValue(Promise.resolve(user));
      requestSignOutOtherDevices.mockReturnValue(() => Promise.resolve());
      loadSecurities.mockReturnValue(() => Promise.resolve());
      loadAWSFields.mockReturnValue(() => Promise.resolve());
      requestVerifyField.mockReturnValue(() => Promise.resolve());

      const result = await store.dispatch(requestLogin(loginPayload));

      const actions = store.getActions();

      expect(actions).toEqual(expectedActions);
      expect(result).toEqual(false);
      expect(Auth.signIn).toBeCalledWith(loginPayload.email, loginPayload.password);
      expect(requestSignOutOtherDevices).toBeCalledWith(signOutPayload);
      expect(loadSecurities).toBeCalled();
      expect(loadAWSFields).toBeCalled();
      expect(requestVerifyField).not.toBeCalled();
    });
  });
});
