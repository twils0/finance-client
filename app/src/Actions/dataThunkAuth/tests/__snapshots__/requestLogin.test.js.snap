// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataThunkAuth requestLogin fails and throws an error when missing 'email' key in payload 1`] = `[Error: Please enter a value for the 'email' key - {}]`;

exports[`dataThunkAuth requestLogin fails and throws an error when missing 'password' key in payload 1`] = `[Error: Please enter a value for the 'password' key - {"email":"testEmail"}]`;
